src/establish-metamask-repository.ts

```diff
@@ -0,0 +1,231 @@
+  if (resolvedRepository.exists) {
+    log('Repository already exists at', resolvedRepository.directoryPath);
+    defaultBranchName = await getDefaultBranchName(
+      resolvedRepository.directoryPath,
+    );
+    lastFetchedDate = await getLastFetchedDate(
+      resolvedRepository.directoryPath,
+    );
+  } else {
+    ({ defaultBranchName, lastFetchedDate } = await cloneRepository({
+      repositoryShortname: resolvedRepository.shortname,
+      repositoryDirectoryPath: resolvedRepository.directoryPath,
+      cachedRepositoriesDirectoryPath,
+      outputLogger,
+    }));
+  }
+
+  return {
+    defaultBranchName,
+    lastFetchedDate,
+    shortname: resolvedRepository.shortname,
+    directoryPath: resolvedRepository.directoryPath,
+    createdAutomatically: resolvedRepository.createdAutomatically,
+  };
+}
+
+/**
+ * Clones a MetaMask repository using the `gh` tool.
+ *
+ * @param args - The arguments to this function.
+ * @param args.repositoryShortname - The name of the repository minus the
+ * organization (so, "utils" instead of "MetaMask/utils").
+ * @param args.repositoryDirectoryPath - The path where the repository should be
+ * cloned to.
+ * @param args.cachedRepositoriesDirectoryPath - The parent directory in which
+ * to keep the new repository.
+ * @param args.outputLogger - Writable streams for output messages.
+ * @returns Information about the cloned repository.
+ * @throws If `repositoryDirectoryPath` is not within
+ * `cachedRepositoriesDirectoryPath`, or you do not have `gh` installed.
+ */
+async function cloneRepository({
+  repositoryShortname,
+  repositoryDirectoryPath,
+  cachedRepositoriesDirectoryPath,
+  outputLogger,
+}: {
+  repositoryShortname: string;
+  repositoryDirectoryPath: string;
+  cachedRepositoriesDirectoryPath: string;
+  outputLogger: AbstractOutputLogger;
+}) {
+  /* istanbul ignore next: There's no way to reproduce this; this is just to be absolutely sure */
+  if (
+    path.dirname(repositoryDirectoryPath) !== cachedRepositoriesDirectoryPath
+  ) {
+    throw new Error(
+      'You seem to be pointing to a directory outside the cached repositories directory. Refusing to proceed to avoid data loss.',
+    );
+  }
+
+  log('Assuming', repositoryShortname, 'is the name of a MetaMask repo');
+
+  log('Removing existing', repositoryDirectoryPath);
+  await fs.promises.rm(repositoryDirectoryPath, {
+    recursive: true,
+    force: true,
+  });
+
+  log('Cloning', repositoryShortname, 'to', repositoryDirectoryPath);
+  outputLogger.logToStdout(
+    `Cloning repository MetaMask/${repositoryShortname}, please wait...`,
+  );
+  await ensureDirectoryStructureExists(cachedRepositoriesDirectoryPath);
+  // NOTE: This requires that you have `gh` installed locally
+  await execa('gh', [
+    'repo',
+    'clone',
+    `MetaMask/${repositoryShortname}`,
+    repositoryDirectoryPath,
+  ]);
+  const defaultBranchName = await getCurrentBranchName(repositoryDirectoryPath);
```

Comment:
My thought here was that in this case, we shouldn't have to make an API request to ask for the default branch, since we just cloned the repo, so we can assume that whatever branch we're on right after cloning _is_ the default branch. Thoughts?

Reply:
Ohhhh I see. The mismatch between this and the variable name made me think it was a mistake, but that does seem like a sensible approach.

Nit: It would be nice to explain here in a comment that we're making this assumption to avoid making an unnecessary network call

Reply:
Sure, makes sense: [d444251](https://github.com/MetaMask/module-lint/pull/30/commits/d444251ac13aca749dbceb58c860ef7b187a7498)